include/mqt-core/na/NAComputation.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include "na/NADefinitions.hpp"
+#include "na/operations/NAOperation.hpp"
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace na {
+class NAComputation {
+protected:
+  std::vector<std::shared_ptr<Point>> initialPositions;
+  std::vector<std::unique_ptr<NAOperation>> operations;
+
+public:
+  NAComputation() = default;
+  NAComputation(NAComputation&& qc) noexcept = default;
+  NAComputation& operator=(NAComputation&& qc) noexcept = default;
+  NAComputation(const NAComputation& qc)
+      : initialPositions(qc.initialPositions) {
+    operations.reserve(qc.operations.size());
+    std::transform(qc.operations.cbegin(), qc.operations.cend(),
+                   std::back_inserter(operations),
+                   [](const auto& op) { return op->clone(); });
+  }
+  NAComputation& operator=(const NAComputation& qc) {
+    if (this != &qc) {
+      initialPositions = qc.initialPositions;
+      operations.clear();
+      operations.reserve(qc.operations.size());
+      std::transform(qc.operations.cbegin(), qc.operations.cend(),
+                     std::back_inserter(operations),
+                     [](const auto& op) { return op->clone(); });
+    }
+    return *this;
+  }
+  virtual ~NAComputation() = default;
+  template <class T> auto emplaceBack(std::unique_ptr<T>&& op) -> void {
+    static_assert(std::is_base_of<NAOperation, T>::value,
+                  "T must be a subclass of NAOperation.");
+    operations.emplace_back(std::move(op));
+  }
+  template <class T> auto emplaceBack(const std::unique_ptr<T>& op) -> void {
+    static_assert(std::is_base_of<NAOperation, T>::value,
+                  "T must be a subclass of NAOperation.");
+    operations.emplace_back(std::move(op));
+  }
+  template <class T, class... Args> auto emplaceBack(Args&&... args) -> void {
+    static_assert(std::is_base_of<NAOperation, T>::value,
```

Comment:
```suggestion
    static_assert(std::is_base_of_v<NAOperation, T>,
                  "T must be a subclass of NAOperation.");
    operations.emplace_back(std::move(op));
  }
  template <class T> auto emplaceBack(const std::unique_ptr<T>& op) -> void {
    static_assert(std::is_base_of_v<NAOperation, T>,
                  "T must be a subclass of NAOperation.");
    operations.emplace_back(std::move(op));
  }
  template <class T, class... Args> auto emplaceBack(Args&&... args) -> void {
    static_assert(std::is_base_of_v<NAOperation, T>,
```

just a little more idiomatic